### PR TITLE
Add extend entry-size calculation (resolved #1387)

### LINKED ISSUE #1387: Discrepancy for calculating size of entries for history truncation
While trying to implement a small feature I discovered a small discrepancy regarding the calculation of the size of an entry between KeePassXC and KeePass2.

## Expected Behavior
The size of entries should be calculated the same way between both applications. Best case would be, that both calculations yield the exact same result.

## Current Behavior
Certain parts are ignored for calculating the size. Tags and AutoTypeAssociations are missing completely, EntryAttributes does only consider values, and ignores keys.

## Possible Solution
Introduce missing calculation parts.

## Debug Info
Revision: 243edda7f2407f600df102f9780b76317e71eedb

### PR DESCRIPTION
**Pull request by ckieschnick**
Extended the calculation for the size of history items to match KeePass2
Small refactoring regarding readability in EntryAttachements. Fixed issue with AutoTypeAssociations which were not pushed to history.

Extended the calculation of the size for entries in the history to match the algorithm in KeePass2

## Description
Added calculation of the size of AutoTypeAssociations and Tags, includes the keys of EntryAttributes. Refactored EntryAttachements to match the general pattern for size calculation. AutoTypeAssociations are now copied to temp-history items to ensure that changes would be pushed to history correctly.

## Motivation and context
Please see #1387.

## How has this been tested?
Changes tested due to unit tests

## Types of changes
- ✅ Bug fix (non-breaking change which fixes an issue)

## Checklist:
- ✅ I have read the **CONTRIBUTING** document. **[REQUIRED]**
- ✅ My code follows the code style of this project. **[REQUIRED]**
- ✅ All new and existing tests passed. **[REQUIRED]**
- ✅ I have compiled and verified my code with `-DWITH_ASAN=ON`. **[REQUIRED]**
- ✅ My change requires a change to the documentation and I have updated it accordingly.
- ✅ I have added tests to cover my changes.

### CODE REVIEW

**Comment by droidmonkey** [comment]
I don't see where this size is shown in the GUI anywhere. Is that something we should add to the properties category of the entry edit? Perhaps the detail view as well.

File: src/core/Entry.cpp [hunk 1]
```diff
@@ -609,12 +609,12 @@ void Entry::truncateHistory()
             // don't calculate size if it's already above the maximum
             if (size <= histMaxSize) {
                 size += historyItem->attributes()->attributesSize();
-
-                const QSet<QByteArray> newAttachments = historyItem->attachments()->values().toSet() - foundAttachments;
-                for (const QByteArray& attachment : newAttachments) {
-                    size += attachment.size();
+                size += historyItem->autoTypeAssociations()->associationsSize();
+                size += historyItem->attachments()->attachmentsSize(foundAttachments);
+                foreach( const QString &tag, historyItem->tags().split(QRegExp(",|;|:"), QString::SkipEmptyParts)){
```
Comment (TheZ3ro) [inline src/core/Entry.cpp]: Can you split this line in 2?  It isn't very clear

Also, can you use C++11 range `for` instead of `foreach`?

File: src/core/AutoTypeAssociations.cpp [hunk 2]
```diff
@@ -103,6 +103,15 @@ int AutoTypeAssociations::size() const
     return m_associations.size();
 }
 
+int AutoTypeAssociations::associationsSize() const
+{
+    int size = 0;
+    foreach (const Association &association, m_associations) {
```
Comment (phoerious) [inline src/core/AutoTypeAssociations.cpp]: Use C++11 range-based for loop (cast m_associations to const before).

File: src/core/EntryAttachments.cpp [hunk 3]
```diff
@@ -34,9 +35,9 @@ bool EntryAttachments::hasKey(const QString& key) const
     return m_attachments.contains(key);
 }
 
-QList<QByteArray> EntryAttachments::values() const
+QSet<QByteArray> EntryAttachments::values() const
```
Comment (phoerious) [inline src/core/EntryAttachments.cpp]: I was wondering why this change is necessary, but then I saw that the only two usages of this method also convert the list to a set, so I'm ok with this change.

File: src/core/EntryAttachments.cpp [hunk 4]
```diff
@@ -151,3 +152,17 @@ bool EntryAttachments::operator!=(const EntryAttachments& other) const
 {
     return m_attachments != other.m_attachments;
 }
+
+int EntryAttachments::attachmentsSize(const QSet<QByteArray> &ignoredAttachments) const
```
Comment (phoerious) [inline src/core/EntryAttachments.cpp]: & belongs to the type, not the parameter name.

File: src/core/EntryAttachments.cpp [hunk 5]
```diff
@@ -151,3 +152,17 @@ bool EntryAttachments::operator!=(const EntryAttachments& other) const
...
+{
+    int size = 0;
+
+    QMapIterator<QString, QByteArray> i(m_attachments);
+    while (i.hasNext()) {
+        i.next();
+        if( ! ignoredAttachments.contains( i.value() )){
```
Comment (phoerious) [inline src/core/EntryAttachments.cpp]: Add space after if and before {, remove spaces around !
Also no spaces around i.value()

File: src/core/EntryAttachments.cpp [hunk 6]
```diff
@@ -151,3 +152,17 @@ bool EntryAttachments::operator!=(const EntryAttachments& other) const
 {
     return m_attachments != other.m_attachments;
 }
+
+int EntryAttachments::attachmentsSize(const QSet<QByteArray> &ignoredAttachments) const
+{
+    int size = 0;
+
+    QMapIterator<QString, QByteArray> i(m_attachments);
```
Comment (phoerious) [inline src/core/EntryAttachments.cpp]: You could use STL-style iterators instead.

File: tests/TestModified.cpp [hunk 7]
```diff
@@ -465,5 +465,52 @@ void TestModified::testHistoryItem()
     entry3->endUpdate();
     QCOMPARE(entry3->historyItems().size(), 2);
 
+    Entry* entry4 = new Entry();
+    entry4->setGroup(root);
+    QCOMPARE(entry4->historyItems().size(), 0);
+
+    int reservedSize = entry4->attributes()->attributesSize();
+    entry4->beginUpdate();
+    entry4->attachments()->set("test1", QByteArray(17000 - 5 - reservedSize + 1, 'a'));
```
Comment (phoerious) [inline tests/TestModified.cpp]: Where do these magic numbers come from?

File: src/core/Entry.cpp (hunk 1 above)
Comment (phoerious) [reply src/core/Entry.cpp]: Also use QRegularExpression, not QRegExp.

**Comment by phoerious** [comment]
@droidmonkey The size only seems to be useful for max history size / items calculation.

**Comment by mstarke** [comment]
@phoerious this is the reason to try to align the size calculations. If all clients divert in this it might result to unexpected behaviour (e.g. missing history entries after Keepass has maintained the file)

**Comment by phoerious** [comment]
@mstarke Sure, no problem with that. Downside is: a change will introduce inconsistencies between KPXC versions. But I think that's a trade-off we can take.

File: src/core/AutoTypeAssociations.cpp [hunk 8]
```diff
@@ -106,7 +106,7 @@ int AutoTypeAssociations::size() const
 int AutoTypeAssociations::associationsSize() const
 {
     int size = 0;
-    foreach (const Association &association, m_associations) {
+    for (const Association &association : m_associations) {
```
Comment (phoerious) [inline src/core/AutoTypeAssociations.cpp]: & belongs to the type. You also need to cast m_associations to const using asConst() from core/global.h. Otherwise the container will detach and you will create unnecessary copies.

File: tests/TestModified.cpp [hunk 9]
```diff
@@ -469,19 +471,21 @@ void TestModified::testHistoryItem()
...
     QCOMPARE(entry4->historyItems().size(), 0);
 
+    const QString key("test");
+
     int reservedSize = entry4->attributes()->attributesSize();
     entry4->beginUpdate();
-    entry4->attachments()->set("test1", QByteArray(17000 - 5 - reservedSize + 1, 'a'));
+    entry4->attachments()->set(key, QByteArray(historyMaxSize - key.size() - reservedSize + 1, 'a'));
     entry4->endUpdate();
     QCOMPARE(entry4->historyItems().size(), 1);
```
Comment (phoerious) [inline tests/TestModified.cpp]: I'm still a little confused here. Aren't you just repeating the above tests with different attachments? I think you actually want to test historySize, not historyItems.

File: src/core/AutoTypeAssociations.cpp (hunk 8 above)
Comment (ckieschnick) [reply src/core/AutoTypeAssociations.cpp]: I'm not quite sure the `asConst` is required at this point since the method itself is annotated with const making (as far as I know) m_associations const as well. I'll look into this and adjust the code accordingly.

File: tests/TestModified.cpp (hunk 9 above)
Comment (ckieschnick) [reply tests/TestModified.cpp]: The tests ensure that all types of associations, attachments, attributes and auto-types (which were not stored to the history at all) with their corresponding keys are considered for the size calculation and prevents regressions in this regard. Maybe the cases should be moved to a separate test case, but the changes were to little to justify this in my eyes.

File: src/core/AutoTypeAssociations.cpp (hunk 8 above)
Comment (phoerious) [reply src/core/AutoTypeAssociations.cpp]: True, discard my comment.

File: tests/TestModified.cpp (hunk 9 above)
Comment (phoerious) [reply tests/TestModified.cpp]: And that's good. I agree they are not entirely redundant. However, I don't see you testing that an entry with an attachment of n bytes actually reports back n + x bytes as size or that adding another attachment/auto-type association removes the oldest history item. It would be amazing if you could add a few quick tests for that as well.

File: tests/TestModified.cpp [hunk 10]
```diff
@@ -417,53 +418,167 @@ void TestModified::testHistoryItem()
     QCOMPARE(entry2->historyItems().size(), 2);
 
     entry2->beginUpdate();
-    entry2->attachments()->remove("test");
+    entry2->attachments()->remove(key);
     entry2->endUpdate();
-    QCOMPARE(entry2->historyItems().size(), 0);
+    QCOMPARE(entry2->attachments()->attachmentsSize(), 0);
+    QCOMPARE(entry2->historyItems().size(), 1);
```
Comment (TheZ3ro) [inline tests/TestModified.cpp]: Why this change for the historyItems size?

Comment (phoerious) [reply tests/TestModified.cpp]: Because the previous attachment gets copied into the history exceeding the configured maximum size.

Comment (TheZ3ro) [reply tests/TestModified.cpp]: Mmmh ok seems reasonable
